### DISCUSSION TITLE
Allow adding cache tags that depend on the response of `fetch` or `unstable_cache`

### DISCUSSION THREAD

**Discussion by romeovs** [vercel/next.js#65521] [Ideas]
### Goals

In some cases it might be useful to provide cache tags to a cache entry, _after_ the result has been received.

For instance: I might fetch a product from the api:

```js
await unstable_cache(() => api.products.get({ id }), ["product", id], {
  tags: [`product:${id}`],
})
// { type: "product", id: 42, ... }
```

however, I might want to include related products in the response:
```js
await unstable_cache(() => api.products.get({ id }, { related: true }), ["product", id], {
  tags: [`product:${id}`],
})
// { type: "product", id: 42, related: [{ type: "product", id: 85 }, ...], ... }
```

It might be useful to be able to tag the response with the ids of those products, even though we only know those _after_ we've received the response. That way, when a related product changes, it can correctly invalidate this response.

Something similar might also be useful for `revalidate`.

### Non-Goals

1. listing current cache tags
2. changing the API in a non-backwards-compatible way

### Background

One example where this is used is [RTK Query](https://redux-toolkit.js.org/rtk-query/usage/automated-refetching), where [`providesTags`](https://redux-toolkit.js.org/rtk-query/api/createApi#providestags) is allowed to be a function that takes the response.

In next something similar could look like:

```js
await unstable_cache(() => api.products.get({ id }, { related: true }), ["product", id], {
  tags(product) {
    return [`product:${id}`, ...product.related.map(p => `product:${p.id}`)]
  }
})
// { type: "product", id: 42, related: [{ type: "product", id: 85 }, ...], ... }
```

### Proposal

Allow passing an (async) function to `tags` that gets passed the response and returns a list of tags.

**Comment by romeovs** [comment]
It does seem to be possible to add cache tags to an existing cache key (without invoking the cached function again).

```ts
function async cachedFn(...args) {
  function fn() {
    // expensive or slow operation, closes over args
  }

  const key = calculateKeyFromArgs(args) // this should be stable

  const res = await unstable_cache(fn, key)
 
  const tags = calculateTagsFromResult(res)

  // A second call does not call fn again, but because the key is the same
  // this will add the tags to the existing cache entry.
  return await unstable_cache(fn, key, {
    tags,
  })
}
```

Can anyone confirm this is the right way of doing this.

I can see one glaring issue with this: revalidating a cache tag using `revalidateTag` between the first and second call to `unstable_cache` in the above function will cause the cache entry to not be cleared properly.

**Reply by PierreCavalet** [reply]
I've open another discussion which is not exactly the same topic but might interest you as a workaround here: https://github.com/vercel/next.js/discussions/75959
